#**Análisis preliminar datos**

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.preprocessing import LabelEncoder,MinMaxScaler,OrdinalEncoder
from statistics import mode


from datetime import datetime, timedelta

Importar los datasets

In [ ]:
df_calendario = pd.read_csv(r"C:\Users\USUARIO\Desktop\CAPSTONE FINAL\Capstone\data_dsmarket\daily_calendar_with_events.csv")

In [ ]:
df_precios = pd.read_csv(r"C:\Users\USUARIO\Desktop\CAPSTONE FINAL\Capstone\data_dsmarket\precios.csv")

In [ ]:
df_ventas = pd.read_csv(r"C:\Users\USUARIO\Desktop\CAPSTONE FINAL\Capstone\data_dsmarket\item_sales.csv")

## **Análisis de Calendario**

In [ ]:
df_calendario.shape


(1913, 5)

In [ ]:
df_calendario.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1913 entries, 0 to 1912
Data columns (total 5 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   date         1913 non-null   object
 1   weekday      1913 non-null   object
 2   weekday_int  1913 non-null   int64 
 3   d            1913 non-null   object
 4   event        26 non-null     object
dtypes: int64(1), object(4)
memory usage: 74.9+ KB


In [ ]:
df_calendario.head(10)

date    weekday  weekday_int     d      event
0  2011-01-29   Saturday            1   d_1        NaN
1  2011-01-30     Sunday            2   d_2        NaN
2  2011-01-31     Monday            3   d_3        NaN
3  2011-02-01    Tuesday            4   d_4        NaN
4  2011-02-02  Wednesday            5   d_5        NaN
5  2011-02-03   Thursday            6   d_6        NaN
6  2011-02-04     Friday            7   d_7        NaN
7  2011-02-05   Saturday            1   d_8        NaN
8  2011-02-06     Sunday            2   d_9  SuperBowl
9  2011-02-07     Monday            3  d_10        NaN

In [ ]:
df_calendario.tail()

date    weekday  weekday_int       d event
1908  2016-04-20  Wednesday            5  d_1909   NaN
1909  2016-04-21   Thursday            6  d_1910   NaN
1910  2016-04-22     Friday            7  d_1911   NaN
1911  2016-04-23   Saturday            1  d_1912   NaN
1912  2016-04-24     Sunday            2  d_1913   NaN

In [ ]:
df_calendario.describe()

weekday_int
count  1913.000000
mean      3.997386
std       2.001175
min       1.000000
25%       2.000000
50%       4.000000
75%       6.000000
max       7.000000

In [ ]:
df_calendario.isnull().sum()

date              0
weekday           0
weekday_int       0
d                 0
event          1887
dtype: int64

In [ ]:
#convertir los valores de "date" en formato "datetime"
df_calendario["date"] = pd.to_datetime(df_calendario["date"])

In [ ]:
df_calendario.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1913 entries, 0 to 1912
Data columns (total 5 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   date         1913 non-null   datetime64[ns]
 1   weekday      1913 non-null   object        
 2   weekday_int  1913 non-null   int64         
 3   d            1913 non-null   object        
 4   event        26 non-null     object        
dtypes: datetime64[ns](1), int64(1), object(3)
memory usage: 74.9+ KB


Características de Calendario:


*   Tiene 1913 filas y 5 columnas
*   Solo la columna 'event' tiene nulos, lo que tiene sentido dada la descripción de la variable.  Los campos nulos se pueden completar con 'Sin evento'






In [ ]:
df_calendario[~(df_calendario['event'].isnull())]

date    weekday  weekday_int       d           event
8    2011-02-06     Sunday            2     d_9       SuperBowl
184  2011-08-01     Monday            3   d_185  Ramadan starts
299  2011-11-24   Thursday            6   d_300    Thanksgiving
337  2012-01-01     Sunday            2   d_338         NewYear
372  2012-02-05     Sunday            2   d_373       SuperBowl
435  2012-04-08     Sunday            2   d_436          Easter
538  2012-07-20     Friday            7   d_539  Ramadan starts
663  2012-11-22   Thursday            6   d_664    Thanksgiving
703  2013-01-01    Tuesday            4   d_704         NewYear
736  2013-02-03     Sunday            2   d_737       SuperBowl
792  2013-03-31     Sunday            2   d_793          Easter
892  2013-07-09    Tuesday            4   d_893  Ramadan starts
1034 2013-11-28   Thursday            6  d_1035    Thanksgiving
1068 2014-01-01  Wednesday            5  d_1069         NewYear
1100 2014-02-02     Sunday            2  d_1101       SuperBowl
1177 2014-04-20     Sunday            2  d_1178          Easter
1247 2014-06-29     Sunday            2  d_1248  Ramadan starts
1398 2014-11-27   Thursday            6  d_1399    Thanksgiving
1433 2015-01-01   Thursday            6  d_1434         NewYear
1464 2015-02-01     Sunday            2  d_1465       SuperBowl
1527 2015-04-05     Sunday            2  d_1528          Easter
1601 2015-06-18   Thursday            6  d_1602  Ramadan starts
1762 2015-11-26   Thursday            6  d_1763    Thanksgiving
1798 2016-01-01     Friday            7  d_1799         NewYear
1835 2016-02-07     Sunday            2  d_1836       SuperBowl
1884 2016-03-27     Sunday            2  d_1885          Easter

In [ ]:
df_calendario['event'].value_counts()

event
SuperBowl         6
Ramadan starts    5
Thanksgiving      5
NewYear           5
Easter            5
Name: count, dtype: int64

## **Análisis de Precios**

In [ ]:
df_precios.shape

(6721786, 6)

In [ ]:
df_precios.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6721786 entries, 0 to 6721785
Data columns (total 6 columns):
 #   Column      Dtype  
---  ------      -----  
 0   Unnamed: 0  int64  
 1   item        object 
 2   category    object 
 3   store_code  object 
 4   yearweek    int64  
 5   sell_price  float64
dtypes: float64(1), int64(2), object(3)
memory usage: 307.7+ MB


In [ ]:
df_precios.head()

Unnamed: 0              item    category store_code  yearweek  sell_price
0           0  ACCESORIES_1_001  ACCESORIES      NYC_1    201328     12.7414
1           1  ACCESORIES_1_001  ACCESORIES      NYC_1    201329     12.7414
2           2  ACCESORIES_1_001  ACCESORIES      NYC_1    201330     10.9858
3           3  ACCESORIES_1_001  ACCESORIES      NYC_1    201331     10.9858
4           4  ACCESORIES_1_001  ACCESORIES      NYC_1    201332     10.9858

In [ ]:
df_precios.describe()

Unnamed: 0      yearweek    sell_price
count  6.721786e+06  6.721786e+06  6.721786e+06
mean   3.483088e+06  2.013824e+05  5.515109e+00
std    2.010729e+06  1.450189e+02  4.381719e+00
min    0.000000e+00  2.011050e+05  1.200000e-02
25%    1.742470e+06  2.012480e+05  2.620100e+00
50%    3.483268e+06  2.014100e+05  4.200000e+00
75%    5.223699e+06  2.015150e+05  7.176000e+00
max    6.965697e+06  2.016170e+05  1.341500e+02

In [ ]:
df_precios.isnull().sum()

Unnamed: 0    0
item          0
category      0
store_code    0
yearweek      0
sell_price    0
dtype: int64

In [ ]:
df_precios[(df_precios['yearweek'].isnull())]


Empty DataFrame
Columns: [Unnamed: 0, item, category, store_code, yearweek, sell_price]
Index: []

In [ ]:
243920/6965706*100

3.5017268888465862

Precios tiene las siguientes caraterísticas generales:


*   Número de filas 6965706 y 5 columnas
*   Solo 'yearweek' tiene elementos nulos y son 243920 ocurrencias, lo que representa un 3,5% de las filas del dataset

In [ ]:
df_precios['item'].value_counts()

item
SUPERMARKET_3_587      2790
HOME_&_GARDEN_1_177    2790
HOME_&_GARDEN_2_283    2790
ACCESORIES_1_337       2790
SUPERMARKET_1_032      2790
                       ... 
HOME_&_GARDEN_1_308     572
HOME_&_GARDEN_1_159     553
HOME_&_GARDEN_1_242     530
SUPERMARKET_3_296       522
SUPERMARKET_2_379       463
Name: count, Length: 3049, dtype: int64

Existen 3049 items diferentes

Analizamos las filas duplicadas

In [ ]:
len(df_precios[df_precios.duplicated()])

0

In [ ]:
df_precios[df_precios.duplicated()].isnull().sum()

Unnamed: 0    0.0
item          0.0
category      0.0
store_code    0.0
yearweek      0.0
sell_price    0.0
dtype: float64

Existen 212120 registros duplicados, y todos ellos tienen la columna yearweek nula.
Se borrarán todos los registros duplicados.

In [ ]:
df_precios.drop_duplicates(inplace=True, keep=False)

In [ ]:
len(df_precios[df_precios.duplicated()])

0

In [ ]:
df_precios.isnull().sum()

Unnamed: 0    0
item          0
category      0
store_code    0
yearweek      0
sell_price    0
dtype: int64

Quedan 425 filas con el valor de yearweek nulo.  Como esa es una variable básica para asignar el precio a un item, se entiende que estos registros no podrán ser utilizados, así que se eliminarán por completo

In [ ]:
df_precios.dropna(inplace=True)

In [ ]:
df_precios.isnull().sum()

Unnamed: 0    0
item          0
category      0
store_code    0
yearweek      0
sell_price    0
dtype: int64

**¿Cómo obtenemos la variable "yearweek"?**

Premisas:
  - El menor valor de "yearweek" en "item_prices.csv" es "201105" y el mayor es "201617"
  - El primer día registrado en "daily_calendar_with_events.csv" es el "d-1" que es un día "sábado" y tiene asignado como "weekday_int" el día "1".  Además indica que es el día 29/01/2011.

Conclusiones:  
Observando la distribución de los datos de "yearweek", "weekday", "weekday_int" y "day" exitentes en "daily_calendar_with_events.csv" y "item_prices.csv", hemos determinado que el cálculo del yearweek no utiliza el estándar que se puede obtener a través del isocalendar, sino que sigue las siguientes reglas:
   - la semana empieza en sábado, que es el día 1 de la semana y el último el viernes, que es el día 7 de la semana
   - la semana 1 de cualquier año empieza a partir del 1er sábado del año, los días anteriores del año pertenecen a la semana 0 del año

Basados en estas premisas incluiremos una columna "yearweek" al dataframe "df_calendario" calculada para adecuarse a las reglas precedentes

In [ ]:
# Definición de función para obtener la weekyear dada un fecha de acuerdo a las premisas anteriores

def a_yearweek(fecha):

  # inicializa una variable con el primer día del año de 'fecha'
  primer_dia_any = datetime(fecha.year,1,1)

  # obtiene el día de la semana del primer día del año (0 es lunes, 6 es domingo)
  dia_sem_primer_dia_any = primer_dia_any.weekday()

  #ajuste para adaptar el inicio de semana de lunes a sabado
  ajuste_sabado = (dia_sem_primer_dia_any + 2) % 7

  # determina el número correspondiente a la primera semana del año
  if dia_sem_primer_dia_any != 5:  # si el primer día del año no es sábado
      primera_semana = 0
  else:
    primera_semana = 1

  # calcula la cantidad de días que han transcurrido desde el inicio del año
  diferencia_dias = (fecha - primer_dia_any).days + ajuste_sabado

  # determina la cantidad de semanas desde el inicio del año
  numero_semana = primera_semana + (diferencia_dias  // 7)


  #retorna la respuesta en el formato yearweek
  return fecha.year * 100 + numero_semana



In [ ]:
df_calendario["yearweek"] = df_calendario['date'].apply(lambda x: a_yearweek(x))

In [ ]:
df_calendario.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1913 entries, 0 to 1912
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   date         1913 non-null   datetime64[ns]
 1   weekday      1913 non-null   object        
 2   weekday_int  1913 non-null   int64         
 3   d            1913 non-null   object        
 4   event        26 non-null     object        
 5   yearweek     1913 non-null   int64         
dtypes: datetime64[ns](1), int64(2), object(3)
memory usage: 89.8+ KB


In [ ]:
df_calendario.head()

date    weekday  weekday_int    d event  yearweek
0 2011-01-29   Saturday            1  d_1   NaN    201105
1 2011-01-30     Sunday            2  d_2   NaN    201105
2 2011-01-31     Monday            3  d_3   NaN    201105
3 2011-02-01    Tuesday            4  d_4   NaN    201105
4 2011-02-02  Wednesday            5  d_5   NaN    201105

In [ ]:
df_precios['item'].value_counts()

item
SUPERMARKET_3_587      2790
HOME_&_GARDEN_1_177    2790
HOME_&_GARDEN_2_283    2790
ACCESORIES_1_337       2790
SUPERMARKET_1_032      2790
                       ... 
HOME_&_GARDEN_1_308     572
HOME_&_GARDEN_1_159     553
HOME_&_GARDEN_1_242     530
SUPERMARKET_3_296       522
SUPERMARKET_2_379       463
Name: count, Length: 3049, dtype: int64

## **Análisis de Ventas**

In [ ]:
df_ventas.shape

(30490, 1920)

In [ ]:
df_ventas.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30490 entries, 0 to 30489
Columns: 1920 entries, id to d_1913
dtypes: int64(1913), object(7)
memory usage: 446.6+ MB


In [ ]:
df_ventas.head()


id              item    category    department   
0  ACCESORIES_1_001_NYC_1  ACCESORIES_1_001  ACCESORIES  ACCESORIES_1  \
1  ACCESORIES_1_002_NYC_1  ACCESORIES_1_002  ACCESORIES  ACCESORIES_1   
2  ACCESORIES_1_003_NYC_1  ACCESORIES_1_003  ACCESORIES  ACCESORIES_1   
3  ACCESORIES_1_004_NYC_1  ACCESORIES_1_004  ACCESORIES  ACCESORIES_1   
4  ACCESORIES_1_005_NYC_1  ACCESORIES_1_005  ACCESORIES  ACCESORIES_1   

               store store_code    region  d_1  d_2  d_3  ...  d_1904  d_1905   
0  Greenwich_Village      NYC_1  New York    0    0    0  ...       1       3  \
1  Greenwich_Village      NYC_1  New York    0    0    0  ...       0       0   
2  Greenwich_Village      NYC_1  New York    0    0    0  ...       2       1   
3  Greenwich_Village      NYC_1  New York    0    0    0  ...       1       0   
4  Greenwich_Village      NYC_1  New York    0    0    0  ...       2       1   

   d_1906  d_1907  d_1908  d_1909  d_1910  d_1911  d_1912  d_1913  
0       0       1       1       1       3       0       1       1  
1       0       0       0       1       0       0       0       0  
2       2       1       1       1       0       1       1       1  
3       5       4       1       0       1       3       7       2  
4       1       0       1       1       2       2       2       4  

[5 rows x 1920 columns]

In [ ]:
df_ventas.describe()

d_1           d_2           d_3           d_4           d_5   
count  30490.000000  30490.000000  30490.000000  30490.000000  30490.000000  \
mean       1.070220      1.041292      0.780026      0.833454      0.627944   
std        5.126689      5.365468      3.667454      4.415141      3.379344   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        0.000000      0.000000      0.000000      0.000000      0.000000   
50%        0.000000      0.000000      0.000000      0.000000      0.000000   
75%        0.000000      0.000000      0.000000      0.000000      0.000000   
max      360.000000    436.000000    207.000000    323.000000    296.000000   

                d_6           d_7           d_8           d_9          d_10   
count  30490.000000  30490.000000  30490.000000  30490.000000  30490.000000  \
mean       0.958052      0.918662      1.244080      1.073663      0.838701   
std        4.785947      5.059495      6.617729      5.917204      4.206199   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        0.000000      0.000000      0.000000      0.000000      0.000000   
50%        0.000000      0.000000      0.000000      0.000000      0.000000   
75%        0.000000      0.000000      0.000000      0.000000      0.000000   
max      314.000000    316.000000    370.000000    385.000000    353.000000   

       ...        d_1904        d_1905        d_1906        d_1907   
count  ...  30490.000000  30490.000000  30490.000000  30490.000000  \
mean   ...      1.370581      1.586159      1.693670      1.248245   
std    ...      3.740017      4.097191      4.359809      3.276925   
min    ...      0.000000      0.000000      0.000000      0.000000   
25%    ...      0.000000      0.000000      0.000000      0.000000   
50%    ...      0.000000      0.000000      0.000000      0.000000   
75%    ...      1.000000      2.000000      2.000000      1.000000   
max    ...    129.000000    160.000000    204.000000     98.000000   

             d_1908        d_1909        d_1910        d_1911        d_1912   
count  30490.000000  30490.000000  30490.000000  30490.000000  30490.000000  \
mean       1.232207      1.159167      1.149000      1.328862      1.605838   
std        3.125471      2.876026      2.950364      3.358012      4.089422   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        0.000000      0.000000      0.000000      0.000000      0.000000   
50%        0.000000      0.000000      0.000000      0.000000      0.000000   
75%        1.000000      1.000000      1.000000      1.000000      2.000000   
max      100.000000     88.000000     77.000000    141.000000    171.000000   

             d_1913  
count  30490.000000  
mean       1.633158  
std        3.812248  
min        0.000000  
25%        0.000000  
50%        0.000000  
75%        2.000000  
max      130.000000  

[8 rows x 1913 columns]

In [ ]:
col = df_ventas.columns[:7]
col

Index(['id', 'item', 'category', 'department', 'store', 'store_code',
       'region'],
      dtype='object')

In [ ]:
df_ventas[col].isnull().sum()

id            0
item          0
category      0
department    0
store         0
store_code    0
region        0
dtype: int64

Ventas tiene las siguientes características generales:


*   30490 filas y 1920 columnas
*   Las columnas representan:
    - id: con la unión de los identificadores de producto y tienda separados por un '_'. Tipo string
    - item: que es el id del producto (incluido en la columna 'id'). Tipo string
    - category: categoría del item. Tipo string
    - department: departamento al que pertenece el item en la tienda.  Es diferente para cada tienda. Tipo string
    - store: nombre de la tienda. Tipo string
    - store_code: código de la tienda (incluido en la columna 'id'). Tipo string
    - region: region donde se encuantra la tienda. tipo string
    - d_1, d_2, d_3: son 1913 columnas donde se indica las ventas del item en la tienda cada día durante casi 6 años

Conclusiones:



*   Para poder utilizar la información de este dataframe, es conveniente crear una fila por cada día.  Es decir, en lugar de una columna.  Para evitar la repetición de los campos que no dependen del día el nuevo dataframe debería contener la información mínima necesaria que nos permita identificar las ventas del día.  
*   Como la columna 'id' contiene la información de 'item' y 'store_code' no es necesario mantener las tres columnas, solo 'id' es suficiente
*   La nueva tabla traspuesta tendrá las columnas:
    - 'id': igual a la columna 'id' de df_ventas
    - 'd': identificador de día, que será igual al nombre da la columna d_1, d_2, d_3, ... , d_1913
    - cant: Cantidad de productos vendidos en la tienda identificados con 'id', el día 'd'
*   Se analizará la relación entre el resto de las variables para extraerlos también de df_ventas de una manera más conveniente










Creamos un nuevo dataframe **df_ventas_dia** que contiene las columnas 'id', 'd' y 'cant', donde 'cant' es la cantidad de items vendidos en la tienda según se indica en 'id' en el día 'd'

In [ ]:
#crear un nuevo dataframe con el valor del 'id', el identificador del dia 'd' y la cantidad de items vendidos en esa tienda ese día 'cant'
# extraer el nombre de las columnas 'd_'
columnas = df_ventas.columns[7:]
columnas

Index(['d_1', 'd_2', 'd_3', 'd_4', 'd_5', 'd_6', 'd_7', 'd_8', 'd_9', 'd_10',
       ...
       'd_1904', 'd_1905', 'd_1906', 'd_1907', 'd_1908', 'd_1909', 'd_1910',
       'd_1911', 'd_1912', 'd_1913'],
      dtype='object', length=1913)

In [ ]:
# Usar melt() para transformar el dataframe
df_ventas_dia = df_ventas.melt(id_vars=['item', 'store_code'], value_vars=columnas, var_name='d', value_name='cantidad')


In [ ]:
df_ventas_dia.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 58327370 entries, 0 to 58327369
Data columns (total 4 columns):
 #   Column      Dtype 
---  ------      ----- 
 0   item        object
 1   store_code  object
 2   d           object
 3   cantidad    int64 
dtypes: int64(1), object(3)
memory usage: 1.7+ GB


En vista de que df_ventas_dia requiere más de 1,7 GB de memoria, eliminaremos todas las filas con 'valor' == 0 ya que no agregan información relevante

In [ ]:
df_ventas_dia.head()

item store_code    d  cantidad
0  ACCESORIES_1_001      NYC_1  d_1         0
1  ACCESORIES_1_002      NYC_1  d_1         0
2  ACCESORIES_1_003      NYC_1  d_1         0
3  ACCESORIES_1_004      NYC_1  d_1         0
4  ACCESORIES_1_005      NYC_1  d_1         0

In [ ]:
df_ventas_dia = df_ventas_dia[(df_ventas_dia['cantidad']>0)]

In [ ]:
df_ventas_dia.tail()

item store_code       d  cantidad
58327361  SUPERMARKET_3_819      PHI_3  d_1913         2
58327362  SUPERMARKET_3_820      PHI_3  d_1913         1
58327364  SUPERMARKET_3_822      PHI_3  d_1913         1
58327365  SUPERMARKET_3_823      PHI_3  d_1913         1
58327368  SUPERMARKET_3_826      PHI_3  d_1913         3

In [ ]:
#Agregar a df_ventas_dia la fecha y la yearweek correspondientes al día indicado en 'd'
df_ventas_dia = df_ventas_dia.merge(df_calendario[['d', 'date', 'yearweek']], on='d', how='left')


In [ ]:
df_ventas_dia.head()

item store_code    d  cantidad       date  yearweek
0  ACCESORIES_1_008      NYC_1  d_1        12 2011-01-29    201105
1  ACCESORIES_1_009      NYC_1  d_1         2 2011-01-29    201105
2  ACCESORIES_1_015      NYC_1  d_1         4 2011-01-29    201105
3  ACCESORIES_1_016      NYC_1  d_1         5 2011-01-29    201105
4  ACCESORIES_1_022      NYC_1  d_1         2 2011-01-29    201105

In [ ]:
#convertir 'date' en un datetime
df_ventas_dia["date"] = pd.to_datetime(df_ventas_dia["date"])

In [ ]:
df_ventas_dia.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18550276 entries, 0 to 18550275
Data columns (total 6 columns):
 #   Column      Dtype         
---  ------      -----         
 0   item        object        
 1   store_code  object        
 2   d           object        
 3   cantidad    int64         
 4   date        datetime64[ns]
 5   yearweek    int64         
dtypes: datetime64[ns](1), int64(2), object(3)
memory usage: 849.2+ MB


In [ ]:
df_ventas_dia.head()

item store_code    d  cantidad       date  yearweek
0  ACCESORIES_1_008      NYC_1  d_1        12 2011-01-29    201105
1  ACCESORIES_1_009      NYC_1  d_1         2 2011-01-29    201105
2  ACCESORIES_1_015      NYC_1  d_1         4 2011-01-29    201105
3  ACCESORIES_1_016      NYC_1  d_1         5 2011-01-29    201105
4  ACCESORIES_1_022      NYC_1  d_1         2 2011-01-29    201105

In [ ]:
# ya teniendo la fecha real no es necesario mantener la columna d, así que se borrará
del(df_ventas_dia['d'])

In [ ]:
df_ventas_dia.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18550276 entries, 0 to 18550275
Data columns (total 5 columns):
 #   Column      Dtype         
---  ------      -----         
 0   item        object        
 1   store_code  object        
 2   cantidad    int64         
 3   date        datetime64[ns]
 4   yearweek    int64         
dtypes: datetime64[ns](1), int64(2), object(2)
memory usage: 707.6+ MB


Hemos reducido la cantidad de memoria requerida por el dataframe en casi un 70%

In [ ]:
df_precios.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6721786 entries, 0 to 6721785
Data columns (total 6 columns):
 #   Column      Dtype  
---  ------      -----  
 0   Unnamed: 0  int64  
 1   item        object 
 2   category    object 
 3   store_code  object 
 4   yearweek    int64  
 5   sell_price  float64
dtypes: float64(1), int64(2), object(3)
memory usage: 307.7+ MB


Para hacer coincidir los tipos de datos de las columnas clave, cambiaremos el tipo de datos de 'yearweek' de df_precios a UInt

In [ ]:
df_ventas_dia['yearweek'].dtype

dtype('int64')

In [ ]:
df_precios['yearweek'] = df_precios['yearweek'].astype(pd.UInt32Dtype())

In [ ]:
df_precios.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6721786 entries, 0 to 6721785
Data columns (total 6 columns):
 #   Column      Dtype  
---  ------      -----  
 0   Unnamed: 0  int64  
 1   item        object 
 2   category    object 
 3   store_code  object 
 4   yearweek    UInt32 
 5   sell_price  float64
dtypes: UInt32(1), float64(1), int64(1), object(3)
memory usage: 288.5+ MB


In [ ]:
df_ventas_dia = pd.merge(df_ventas_dia, df_precios[['item', 'store_code', 'yearweek', 'sell_price']], on=['item', 'store_code', 'yearweek'], how='left')

In [ ]:
df_ventas_dia.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18550276 entries, 0 to 18550275
Data columns (total 6 columns):
 #   Column      Dtype         
---  ------      -----         
 0   item        object        
 1   store_code  object        
 2   cantidad    int64         
 3   date        datetime64[ns]
 4   yearweek    int64         
 5   sell_price  float64       
dtypes: datetime64[ns](1), float64(1), int64(2), object(2)
memory usage: 849.2+ MB


In [ ]:
df_ventas_dia.isnull().sum()

item              0
store_code        0
cantidad          0
date              0
yearweek          0
sell_price    18744
dtype: int64

In [ ]:
df_ventas_dia[((df_ventas_dia["sell_price"].isnull()) & (df_ventas_dia["yearweek"] != 201153) & (df_ventas_dia["yearweek"] != 201252))]

item store_code  cantidad       date  yearweek   
2280219    SUPERMARKET_3_315      NYC_1         3 2012-01-01    201200  \
2280663  HOME_&_GARDEN_2_149      NYC_2         1 2012-01-01    201200   
2280737  HOME_&_GARDEN_2_458      NYC_2         1 2012-01-01    201200   
2280870    SUPERMARKET_3_149      NYC_2         1 2012-01-01    201200   
2281102     ACCESORIES_1_153      NYC_3         4 2012-01-01    201200   
...                      ...        ...       ...        ...       ...   
5289165  HOME_&_GARDEN_1_081      BOS_2         2 2012-12-28    201251   
5289274  HOME_&_GARDEN_1_355      BOS_2         3 2012-12-28    201251   
5290315  HOME_&_GARDEN_2_451      BOS_3         1 2012-12-28    201251   
5290457    SUPERMARKET_2_261      BOS_3         1 2012-12-28    201251   
5291309    SUPERMARKET_2_202      PHI_1         1 2012-12-28    201251   

         sell_price  
2280219         NaN  
2280663         NaN  
2280737         NaN  
2280870         NaN  
2281102         NaN  
...             ...  
5289165         NaN  
5289274         NaN  
5290315         NaN  
5290457         NaN  
5291309         NaN  

[11209 rows x 6 columns]

In [ ]:
df_ventas_dia.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18550276 entries, 0 to 18550275
Data columns (total 7 columns):
 #   Column      Dtype         
---  ------      -----         
 0   item        object        
 1   store_code  object        
 2   cantidad    int64         
 3   date        datetime64[ns]
 4   yearweek    int64         
 5   sell_price  float64       
 6   monto       float64       
dtypes: datetime64[ns](1), float64(2), int64(2), object(2)
memory usage: 990.7+ MB


In [ ]:
df_ventas_dia.head()

item store_code  cantidad       date  yearweek  sell_price   
0  ACCESORIES_1_008      NYC_1        12 2011-01-29    201105      0.6118  \
1  ACCESORIES_1_009      NYC_1         2 2011-01-29    201105      2.0748   
2  ACCESORIES_1_015      NYC_1         4 2011-01-29    201105      0.9310   
3  ACCESORIES_1_016      NYC_1         5 2011-01-29    201105      0.9310   
4  ACCESORIES_1_022      NYC_1         2 2011-01-29    201105      9.1238   

     monto  
0   7.3416  
1   4.1496  
2   3.7240  
3   4.6550  
4  18.2476

In [ ]:
df_ventas_dia = df_ventas_dia.sort_values(['item','store_code','date'])

In [ ]:
df_ventas_dia['sell_price'] = df_ventas_dia.groupby(['item','store_code'])['sell_price'].fillna(method='ffill').fillna(method='bfill')

In [ ]:
df_ventas_dia['monto'] = df_ventas_dia['cantidad'] * df_ventas_dia['sell_price']

In [ ]:
df_ventas_dia.isnull().sum()

item          0
store_code    0
cantidad      0
date          0
yearweek      0
sell_price    0
monto         0
dtype: int64

**Variables relacionadas con 'store'**

In [ ]:
len(df_ventas)

30490

In [ ]:
# estudiamos la relación entre las variables relacionadas con la tienda
df_ventas[['store','store_code','region']].value_counts()

store              store_code  region      
Back_Bay           BOS_3       Boston          3049
Brooklyn           NYC_4       New York        3049
Greenwich_Village  NYC_1       New York        3049
Harlem             NYC_2       New York        3049
Midtown_Village    PHI_1       Philadelphia    3049
Queen_Village      PHI_3       Philadelphia    3049
Roxbury            BOS_2       Boston          3049
South_End          BOS_1       Boston          3049
Tribeca            NYC_3       New York        3049
Yorktown           PHI_2       Philadelphia    3049
Name: count, dtype: int64

Hay 10 tiendas en total

Se puede observar que cada 'store_code' se corresponde unequivocamente con una tienda y que cada tienda pertenece a una sola región.  No es necesario almacenar toda esta información en el dataframe df_ventas, se creará un dataframe df_tiendas que contenga esta información.


In [ ]:
#Creamos el df_tiendas
df_tiendas = df_ventas[['store','store_code','region']].copy()

In [ ]:
df_tiendas.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30490 entries, 0 to 30489
Data columns (total 3 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   store       30490 non-null  object
 1   store_code  30490 non-null  object
 2   region      30490 non-null  object
dtypes: object(3)
memory usage: 714.7+ KB


In [ ]:
df_tiendas = df_tiendas.drop_duplicates().reset_index(drop=True)


In [ ]:
df_tiendas

store store_code        region
0  Greenwich_Village      NYC_1      New York
1             Harlem      NYC_2      New York
2            Tribeca      NYC_3      New York
3           Brooklyn      NYC_4      New York
4          South_End      BOS_1        Boston
5            Roxbury      BOS_2        Boston
6           Back_Bay      BOS_3        Boston
7    Midtown_Village      PHI_1  Philadelphia
8           Yorktown      PHI_2  Philadelphia
9      Queen_Village      PHI_3  Philadelphia

In [ ]:
df_tiendas.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10 entries, 0 to 9
Data columns (total 3 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   store       10 non-null     object
 1   store_code  10 non-null     object
 2   region      10 non-null     object
dtypes: object(3)
memory usage: 368.0+ bytes


**Analizamos la relación entre las variables 'item','category', 'department'**



In [ ]:
# Estudiamos la relación entre los'item' y las 'category'
# para ello primero contamos los 'item'
df_ventas['item'].value_counts()

item
ACCESORIES_1_001       10
SUPERMARKET_2_210      10
SUPERMARKET_2_201      10
SUPERMARKET_2_202      10
SUPERMARKET_2_203      10
                       ..
HOME_&_GARDEN_1_464    10
HOME_&_GARDEN_1_465    10
HOME_&_GARDEN_1_466    10
HOME_&_GARDEN_1_467    10
SUPERMARKET_3_827      10
Name: count, Length: 3049, dtype: int64

In [ ]:
# son 3049 'item'
# ahora vamos a validar la relación entre 'item' y category'
df_ventas[['item','category']].value_counts()

item                 category     
ACCESORIES_1_001     ACCESORIES       10
SUPERMARKET_2_210    SUPERMARKET      10
SUPERMARKET_2_201    SUPERMARKET      10
SUPERMARKET_2_202    SUPERMARKET      10
SUPERMARKET_2_203    SUPERMARKET      10
                                      ..
HOME_&_GARDEN_1_464  HOME_&_GARDEN    10
HOME_&_GARDEN_1_465  HOME_&_GARDEN    10
HOME_&_GARDEN_1_466  HOME_&_GARDEN    10
HOME_&_GARDEN_1_467  HOME_&_GARDEN    10
SUPERMARKET_3_827    SUPERMARKET      10
Name: count, Length: 3049, dtype: int64

In [ ]:
df_ventas['category'].value_counts()

category
SUPERMARKET      14370
HOME_&_GARDEN    10470
ACCESORIES        5650
Name: count, dtype: int64

Como hay la misma cantidad de 'item' (3049) que de 'item' y 'category' podemos asumir que cada 'item' pertenece a una 'category'

Entonces, sacaremos esta relación del df_ventas dejando solo la referencia al 'item' y creando un dataframe df_cat_item que contenga la relación  

In [ ]:
# creamos df_cat_item con 'item' y 'category'
df_cat_item = df_ventas[['item','category']].copy()
#eliminamos las filas repetidas y reseteamos los índices
df_cat_item = df_cat_item.drop_duplicates().reset_index(drop=True)

In [ ]:
df_cat_item.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3049 entries, 0 to 3048
Data columns (total 2 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   item      3049 non-null   object
 1   category  3049 non-null   object
dtypes: object(2)
memory usage: 47.8+ KB


Sabemos que el 'departament' de un item viene dado por 'item' y 'store_code'.  Por tanto, podríamos crear un dataframe df_departamento que contenga 'item', 'store_code' y 'department'

In [ ]:
# creamos df_department con 'item' y 'category'
df_department = df_ventas[['item','store_code','department']].copy()
#eliminamos las filas repetidas y reseteamos los índices
df_department = df_department.drop_duplicates().reset_index(drop=True)

In [ ]:
df_department.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30490 entries, 0 to 30489
Data columns (total 3 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   item        30490 non-null  object
 1   store_code  30490 non-null  object
 2   department  30490 non-null  object
dtypes: object(3)
memory usage: 714.7+ KB


In [ ]:
df_department['department'].value_counts()

department
SUPERMARKET_3      8230
HOME_&_GARDEN_1    5320
HOME_&_GARDEN_2    5150
ACCESORIES_1       4160
SUPERMARKET_2      3980
SUPERMARKET_1      2160
ACCESORIES_2       1490
Name: count, dtype: int64

# **Exportar todos los dataframes creados como .csv**

In [ ]:
df_department.to_csv('C:\\Users\\USUARIO\\Desktop\\CAPSTONE FINAL\\Capstone\\data_dsmarket\\department.csv')

In [ ]:
df_precios.to_csv('C:\\Users\\USUARIO\\Desktop\\CAPSTONE FINAL\\Capstone\\data_dsmarket\\precios.csv')

In [ ]:
df_ventas_dia.to_csv('C:\\Users\\USUARIO\\Desktop\\CAPSTONE FINAL\\Capstone\\data_dsmarket\\ventas_dia.csv')

In [ ]:
df_cat_item.to_csv('C:\\Users\\USUARIO\\Desktop\\CAPSTONE FINAL\\Capstone\\data_dsmarket\\categoria_item.csv')

In [ ]:
df_tiendas.to_csv('C:\\Users\\USUARIO\\Desktop\\CAPSTONE FINAL\\Capstone\\data_dsmarket\\tiendas.csv')

In [ ]:
df_calendario.to_csv('C:\\Users\\USUARIO\\Desktop\\CAPSTONE FINAL\\Capstone\\data_dsmarket\\calendario.csv')